### PR TITLE
feat(gatsby-remark-embed-snippet): Add the ability to embed named snippets

### PR DESCRIPTION
**Pull request by BobWall23**
## Description
Adds a feature to the gatsby-remark-embed-snippet plugin that makes it possible for someone to add inline comments to a source file to delineate one or more named snippets, and to embed snippets by name.

### Documentation

The new feature is documented in the `README.md` for the `gatsby-remark-embed-snippet` plugin.

### CODE REVIEW

**Comment by pieh** [comment]
This looks awesome!

I do however would want to discuss a bit way in which we specify snippet names to be used. Doing `embed:file#SNsnippetName` will work but I do wonder if we should try to take hint out of `gatsby-remark-prismjs` book and instead have something like `embed:file{ snippet: "snippetName" }` - for users not familiar with `#SN<snippetName>` syntax I think this will be easier to understand (and also give us room to implement other options/attributes in the future without overloading `#` (which can be reserved for line numbers)

**Comment by pieh** [comment]
Another thing is about marking snippets - using them with comments is fine, i just think we should use conventions we already have for line highlighting - so (potentially) ` // START SNIPPET bar` -> `// start-snippet{bar}` - this is just so API is consistent and parts of it feel like they go well together.

Let me know what do you think

**Comment by BobWall23** [comment]
Thanks for the suggestions - I think it is good to make this fit in better. I didn't look far enough outside this plugin to see how some of the others worked. I will update to match the prismjs syntax better, using the syntax you suggested.

**Comment by pieh** [comment]
Those were just examples - we might end up with them, but I'd like to encourage you to take a peek into what syntax conventions are used currently and maybe see if you can come up with better ones. You could take a look in both in `gatsby-remark-embed-snippet` and `gatsby-remark-prismjs` (because those 2 are pretty close together and both are used for similar results - showing some code blocks).

**Comment by BobWall23** [comment]
Those suggestions do seem to be in line with the syntax used in the `prismjs` plugin. The only option syntax in the `embed-snippet` plugin is the `#L` parsing.

I think it is pretty unfortunate that prismjs used `{optionName : value}` instead of `{"optionName" : value}`, to facilitate standard JSON parsing. I can continue with similar syntax (no double quotes around the option names) or use standard JSON conventions. I prefer the latter, in spite of the slight mismatch in syntax.

**Comment by BobWall23** [comment]
I went with a change that doesn't require double quotes around the `snippet` keyword - it would be a trivial change plus update to the docs to switch over to valid JSON format. I think that would be more robust if new options were added.

I wonder if I blew the commits by not using the conventional commit standard? I was thinking I just needed to worry about the pull request, not the individual commits to my fork. Hopefully that is correct.

**Comment by pieh** [comment]
> I wonder if I blew the commits by not using the conventional commit standard? I was thinking I just needed to worry about the pull request, not the individual commits to my fork. Hopefully that is correct.

This is correct - we are "squashing and merging", so single commit will end up in master - and if there is more than 1 commit in PR - PR title is used

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 1]
```diff
@@ -73,6 +99,25 @@ module.exports = ({ markdownAST, markdownNode }, { directory } = {}) => {
           .split(`\n`)
           .filter((_, lineNumber) => lines.includes(lineNumber + 1))
           .join(`\n`)
+      } else if (sname.length) {
+        let index1 = code.indexOf(`start-snippet{${sname}}`)
+        if (index1 > -1) {
+          let index2 = code.indexOf(`\n`, index1)
+          if (index2 > -1) {
+            index2 = index2 + 1 // skip the newline
```
[the remark this answers is not in the document]
Comment (BobWall23) [reply packages/gatsby-remark-embed-snippet/src/index.js:107]: I didn't want the newline at the end of the comment containing `start-snippet` to be included in the included code.

**Comment by BobWall23** [comment]
I notice that the lint check is failing due to an extra space in `docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md`. If I fix that, the lint passes - should I commit that change into this PR?

File: packages/gatsby-remark-embed-snippet/src/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby-remark-embed-snippet/src/index.js:107]: Ah, right - this makes sense.

I do wonder if we could make the code more concise (and drop `index1` up to `index4` which makes it bit hard to follow) in the process:

what instead of using `(last)indexOf` we would use regular expressions:

```js
const startingSnippetDirectiveMatcher = new RegExp(`start-snippet{${sname}}[^\r\n]*[\r\n](.*)`, `gs`)
const startSnippetMatch = startingSnippetDirectiveMatcher.exec(code)
if (startSnippetMatch?.length >= 2) {
  code =  startSnippetMatch[1]
}

const endingSnippetDirectiveMatcher = new RegExp(`(.*)[\r\n][^\r\n]*end-snippet{${sname}}`, `gs`)
const endSnippetMatch = endingSnippetDirectiveMatcher.exec(code)
if (endSnippetMatch?.length >= 2) {
  code =  endSnippetMatch[1]
}
```

This would pick next lines after `start-snippet{name}` in first step and lines before `end-snippet` in the second.

Alternative is to keep current code - just move it to some utility file to not show all the "gruesome" details of extracting snippets in main file which should be bit more high level than details like.

I'm fine with both approaches

**Comment by pieh** [comment]
> I notice that the lint check is failing due to an extra space in `docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md`. If I fix that, the lint passes - should I commit that change into this PR?

You can merge `master` branch of gatsbyjs/gatsby into your branch and push it (lint was fixed in master already).

Assuming `origin` = `gatsbyjs/gatsby`:
```
git fetch origin master
git merge origin/master
git push <your_fork> packages/add-named-snippets
```

should fix lint problems


Alternatively - you can go ahead and fix the lint separately in your branch - because it was fixed already in master - this change should not show in PR view

File: packages/gatsby-remark-embed-snippet/src/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (BobWall23) [reply packages/gatsby-remark-embed-snippet/src/index.js]: I like the regular expressions - used that approach. Thanks for the suggestion!

**Comment by pieh** [comment]
All right, code has sign off from me - it looks great and tests are great! Now just need to review for docs changes

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 2]
```diff
@@ -51,6 +52,7 @@ module.exports = ({ markdownAST, markdownNode }, { directory } = {}) => {
 
       // Embed specific lines numbers of a file
       let lines = []
+      var sname = ``
```
Comment (pieh) [inline packages/gatsby-remark-embed-snippet/src/index.js:55]: Just small NIT :)

```suggestion
      let sname = ``
```

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 3]
```diff
@@ -61,6 +63,30 @@ module.exports = ({ markdownAST, markdownNode }, { directory } = {}) => {
         }
         // Remove everything after the range prefix from file path
         snippetPath = snippetPath.slice(0, rangePrefixIndex)
+      } else {
+        // Check to see if there is a {snippet: "snippetName"} following the file path.
+        // This syntax could support additional options in the future - for now, only
+        // handle a string that contains a `snippet :` option.
+        var optionIndex = snippetPath.indexOf(`{`)
+        if (optionIndex > -1) {
+          var optionStr = snippetPath.slice(optionIndex)
```
Comment (pieh) [inline packages/gatsby-remark-embed-snippet/src/index.js:72]:
```suggestion
          let optionStr = snippetPath.slice(optionIndex)
```

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 4]
```diff
@@ -61,6 +63,30 @@ module.exports = ({ markdownAST, markdownNode }, { directory } = {}) => {
         }
         // Remove everything after the range prefix from file path
         snippetPath = snippetPath.slice(0, rangePrefixIndex)
+      } else {
+        // Check to see if there is a {snippet: "snippetName"} following the file path.
+        // This syntax could support additional options in the future - for now, only
+        // handle a string that contains a `snippet :` option.
+        var optionIndex = snippetPath.indexOf(`{`)
```
Comment (pieh) [inline packages/gatsby-remark-embed-snippet/src/index.js:70]:
```suggestion
        const optionIndex = snippetPath.indexOf(`{`)
```

File: packages/gatsby-remark-embed-snippet/src/index.js [hunk 5]
```diff
@@ -61,6 +63,30 @@ module.exports = ({ markdownAST, markdownNode }, { directory } = {}) => {
...
+        if (optionIndex > -1) {
+          var optionStr = snippetPath.slice(optionIndex)
+          snippetPath = snippetPath.slice(0, optionIndex)
+          try {
+            var optVal = JSON.parse(
```
Comment (pieh) [inline packages/gatsby-remark-embed-snippet/src/index.js:75]:
```suggestion
            const optVal = JSON.parse(
```

File: packages/gatsby-remark-embed-snippet/README.md [hunk 6]
```diff
@@ -332,3 +270,119 @@ function App() {
...
 }
 ```
+
+### Specifying snippets by name
+
+As an alternative to selecting a range of lines from a file, you can add `start-snippet{snippet-name}` and `end-snippet{snippet-name}` in comments in your files. The inclusion of a name for a snippet allows you to create an example file that contains multiple snippets that you reference from different places.
+
+You specify that you want to only include a named snippet from the embed using the syntax `{snippet: "snippet-name"}`.
+
+**Rust example**:
```
Comment (Ekwuno) [inline packages/gatsby-remark-embed-snippet/README.md:280]: Do you think we can get a JavaScript example of specific snippets by name? Rust is good but most off our users would relate more to a JS code snippets.

File: docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md [hunk 7]
```diff
@@ -71,4 +71,4 @@ It’s truly exciting to look back to see how far we have come in the last five
 
 So no matter what happens over the next five years, there are things that will not alter. Gatsby the open source framework is always going to be open source. Always going to be free, always going to be supported, and always with the community as co-pilot.
 
-_Ready to dive in for even more Gatsby goodness?  Join us at our first-ever virtual Gatsby Days, two half days of speakers, demos, and All Things Gatsby coming up on June 2 & 3rd. Register now: https://www.gatsbyjs.com/virtual-gatsby-days-registration/_
+_Ready to dive in for even more Gatsby goodness? Join us at our first-ever virtual Gatsby Days, two half days of speakers, demos, and All Things Gatsby coming up on June 2 & 3rd. Register now: https://www.gatsbyjs.com/virtual-gatsby-days-registration/_
```
Comment (Ekwuno) [inline docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md:74]: I am not sure if you meant to edit this. 
```suggestion
_Ready to dive in for even more Gatsby goodness?  Join us at our first-ever virtual Gatsby Days, two half days of speakers, demos, and All Things Gatsby coming up on June 2 & 3rd. Register now: https://www.gatsbyjs.com/virtual-gatsby-days-registration/_
```

File: packages/gatsby-remark-embed-snippet/README.md [hunk 8]
```diff
@@ -332,3 +270,119 @@ function App() {
   )
 }
 ```
+
+### Specifying snippets by name
+
+As an alternative to selecting a range of lines from a file, you can add `start-snippet{snippet-name}` and `end-snippet{snippet-name}` in comments in your files. The inclusion of a name for a snippet allows you to create an example file that contains multiple snippets that you reference from different places.
+
+You specify that you want to only include a named snippet from the embed using the syntax `{snippet: "snippet-name"}`.
```
Comment (Ekwuno) [inline packages/gatsby-remark-embed-snippet/README.md:278]:
```suggestion
You can specify that you want to only include a named snippet from the embed, using the syntax `{snippet: "snippet-name"}`.
```

File: packages/gatsby-remark-embed-snippet/README.md (hunk 6 above)
Comment (BobWall23) [reply packages/gatsby-remark-embed-snippet/README.md:280]: I added a Rust example in part to show that `gatsby-remark-embed-snippet` now understands `.rs` as a Rust file extension, but that might have not been obvious.

File: docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md (hunk 7 above)
Comment (BobWall23) [reply docs/blog/2020-05-22-happy-fifth-bday-gatsby/index.md:74]: Yes. The extra space after the `?` breaks lint in CircleCI.

File: packages/gatsby-remark-embed-snippet/README.md [hunk 9]
```diff
@@ -332,3 +270,119 @@ function App() {
...
+    // start-snippet{invokeA}
+    let xfact = factorial(x)
+    // end-snippet{invokeA}
+    println!(`{} factorial is {}`, x, xfact)
+}
+```
+
+Will produce something like this:
+
+```
```
Comment (muescha) [inline packages/gatsby-remark-embed-snippet/README.md:313]: add some type here - for example text or markdown?
````suggestion
```text
````

File: packages/gatsby-remark-embed-snippet/README.md [hunk 10]
```diff
@@ -332,3 +270,119 @@ function App() {
...
+The function to use is
```
Comment (muescha) [inline packages/gatsby-remark-embed-snippet/README.md:314]:
```suggestion
The function to use is:
```

should match the source text...

i miss somehow a `:` at second line `And it is invoked via`

**Comment by AishaBlake** [comment]
Note: this says that we still need a Code owner review from the core team but pieh signed off on this a while back!
